modules/storage-api/src/test/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageTest.java

```diff
@@ -0,0 +1,206 @@
+     */
+    @Test
+    public void testAddWrite() throws Exception {
+        MvPartitionStorage pk = partitionStorage();
+
+        TestKey key = new TestKey(10, "foo");
+        TestValue value = new TestValue(20, "bar");
+
+        BinaryRow binaryRow = binaryRow(key, value);
+
+        pk.addWrite(binaryRow, UUID.randomUUID());
+
+        // Attempt to write from another transaction.
+        assertThrows(TxIdMismatchException.class, () -> pk.addWrite(binaryRow, UUID.randomUUID()));
+
+        // Read without timestamp returns uncommited row.
+        assertEquals(value, value(pk.read(binaryKey(key), null)));
+
+        // Read with timestamp returns null.
+        assertNull(pk.read(binaryKey(key), Timestamp.nextVersion()));
+    }
+
+    /**
+     * Tests basic invariants of {@link MvPartitionStorage#abortWrite(BinaryRow)}.
+     */
+    @Test
+    public void testAbortWrite() throws Exception {
+        MvPartitionStorage pk = partitionStorage();
+
+        TestKey key = new TestKey(10, "foo");
+        TestValue value = new TestValue(20, "bar");
+
+        pk.addWrite(binaryRow(key, value), UUID.randomUUID());
+
+        pk.abortWrite(binaryKey(key));
+
+        // Aborted row can't be read.
+        assertNull(pk.read(binaryKey(key), null));
+    }
+
+    /**
+     * Tests basic invariants of {@link MvPartitionStorage#commitWrite(BinaryRow, Timestamp)}.
+     */
+    @Test
+    public void testCommitWrite() throws Exception {
+        MvPartitionStorage pk = partitionStorage();
+
+        TestKey key = new TestKey(10, "foo");
+        TestValue value = new TestValue(20, "bar");
+
+        BinaryRow binaryRow = binaryRow(key, value);
+
+        pk.addWrite(binaryRow, UUID.randomUUID());
+
+        Timestamp tsBefore = Timestamp.nextVersion();
+
+        Timestamp tsExact = Timestamp.nextVersion();
+        pk.commitWrite(binaryRow, tsExact);
+
+        Timestamp tsAfter = Timestamp.nextVersion();
+
+        // Row is invisible at the time before writing.
+        assertNull(pk.read(binaryRow, tsBefore));
+
+        // Row is valid at the time during and after writing.
+        assertEquals(value, value(pk.read(binaryRow, null)));
+        assertEquals(value, value(pk.read(binaryRow, tsExact)));
+        assertEquals(value, value(pk.read(binaryRow, tsAfter)));
+
+        TestValue newValue = new TestValue(30, "duh");
+
+        pk.addWrite(binaryRow(key, newValue), UUID.randomUUID());
+
+        // Same checks, but now there are two different versions.
+        assertNull(pk.read(binaryRow, tsBefore));
+
+        assertEquals(newValue, value(pk.read(binaryRow, null)));
+
+        assertEquals(value, value(pk.read(binaryRow, tsExact)));
+        assertEquals(value, value(pk.read(binaryRow, tsAfter)));
+        assertEquals(value, value(pk.read(binaryRow, Timestamp.nextVersion())));
+
+        pk.commitWrite(binaryKey(key), Timestamp.nextVersion());
+
+        assertEquals(newValue, value(pk.read(binaryRow, null)));
```

Comment:
should we also check that previous value can still be extracted after a commit?

Reply:
There is a test for that

Reply:
Why not